packages/search/src/util/merge-sort/merge.ts

```diff
@@ -0,0 +1,300 @@
+interface INode<T> {
+  data: T;
+  label: string;
+}
+
+interface IBinaryHeap<T> {
+  insert(node: INode<T>): void;
+  remove(): INode<T> | null;
+}
+
+type comparator<T> = (one: T, two: T) => number;
+
+type HeapDirection = "ASC" | "DESC";
+
+/**
+ * k-way merge implementation that merges (and potentially sorts) k individually sorted arrays
+ * based on the k-way merge algorithm (https://en.wikipedia.org/wiki/K-way_merge_algorithm)
+ *
+ * If a comparator function used to compare data is provided, a k-way merge sort is performed using
+ * a Binary Heap implementation. Otherwise data from each input array is merged incrementally
+ *
+ * The returned object contains merged (and potentially sorted) data, as well as key-value
+ * pairs that represent the index of a result set in the "data" parameter (the key) tied to how
+ * many results (the value) of that result set were added to the final merged data array.
+ * For example, if "data" represents an array of two result sets, with 3 and 5 results added to the final
+ * result set, respectively, the returned object would be { data: [...], "0": 3, "1": 5 }
+ *
+ * @param data An array of of result sets, each an array of type T
+ * @param resultLimit the maximum number of merged results to return, defaults to 10
+ * @param cmptr comparator function that takes in two instances of type T and returns a number to determine sorting order
+ * @param direction specifies whether data should be ordered ascending or descending
+ * @returns list of results and key-value pairs indicating how many from each were added to returned data
+ */
+export function kMerge<T>(
+  data: T[][],
+  resultLimit: number = 10,
+  cmptr?: comparator<T[]>,
+  direction?: HeapDirection
+): { data: T[]; [key: string]: any } {
+  const results = cmptr
+    ? kMergeSort(data, cmptr, direction, resultLimit)
+    : kMergeDefault(data, resultLimit);
+
+  // Zero-fill any result set that did not have any results added to final
+  for (let i = 0; i < data.length; i++) {
```

Comment:
why do we do this?

Reply:
This is completely optional, but if a result set that is provided as input had 0 results added to the final set, the corresponding index key in the return object would be undefined. This sets it to 0 so client doesn't have to check.